File: compiler/src/runner/Closure.ts
import type { Location } from "#utils";
import { WriterError } from "../ast/WriterError.ts";
import type { Frame } from "./Frame.ts";
import type { Variable } from "./Variable.ts";

export class Closure {
  readonly #globals: Frame;
  readonly #frames: Array<Frame>;

  constructor(globals: Frame, frames: Array<Frame>) {
    this.#globals = globals;
    this.#frames = frames;
  }

  search(name: string, displayName: string, location: Location) {
    for (const frame of this.#frames) {
      const possible = frame.search(name);
      if (possible) return possible;
    }

    throw new WriterError(`Variable ${displayName} not resolved`, location);
  }

  withFrame(frame: Frame) {
    return new Closure(this.#globals, [...this.#frames, frame]);
  }

  withVariable(name: string, value: Variable | Promise<Variable>) {
    const frame = this.#frames[this.#frames.length - 1];
    if (!frame) throw new Error("Not within a frame");
    return new Closure(this.#globals, [...this.#frames.filter((f) => f !== frame), frame.withVariable(name, value)]);
  }

  searchGlobal(name: string, location: Location) {
    const possible = this.#globals.search(name);
    if (!possible) throw new WriterError(`Could not find external ${name}`, location);

    return possible;
  }
}

File: runner/src/Storage/Closure.cpp
#include "Closure.h"

namespace Storage {
Closure::Closure(Frame* globals, std::vector<Frame*> frames)
{
  this->globals = globals;
  this->frames = frames;
}

Closure::~Closure()
{
}

const Variable* Closure::search(std::string name)
{
  for (const auto& frame : this->frames) {
    auto possible = frame->search(name);
    if (possible != nullptr) {
      return possible;
    }
  }

  throw "Variable not resolved";
}

Closure* Closure::with_frame(Frame* frame)
{
  auto input = std::vector<Frame*>();
  for (const auto& frame : this->frames) {
    input.push_back(frame);
  }

  input.push_back(frame);

  return new Closure(this->globals, input);
}

Closure* Closure::add_variable(std::string name, const Variable* value)
{
  auto frame = *this->frames.end();
  frame->add_variable(name, value);
  return this;
}

const Variable* Closure::add_temp_variable(const Variable* value)
{
  auto frame = *this->frames.end();
  frame->add_temp_variable(value);
  return value;
}

const Variable* Closure::search_global(std::string name)
{
  auto possible = this->globals->search(name);
  if (possible != nullptr) {
    return possible;
  }

  throw "Global not resolved";
}
}
